lsp/src/Language/LSP/VFS.hs

```diff
@@ -380,103 +381,42 @@ data CodePointRange =
 makeFieldsNoPrefix ''CodePointPosition
 makeFieldsNoPrefix ''CodePointRange
 
-{- Note [Converting between code points and code units]
-This is inherently a somewhat expensive operation, but we take some care to minimize the cost.
-In particular, we use the good asymptotics of 'Rope' to our advantage:
-- We extract the single line that we are interested in in time logarithmic in the number of lines.
-- We then split the line at the given position, and check how long the prefix is, which takes
-linear time in the length of the (single) line.
-
-We also may need to convert the line back and forth between ropes with different indexing. Again
-this is linear time in the length of the line.
-
-So the overall process is logarithmic in the number of lines, and linear in the length of the specific
-line. Which is okay-ish, so long as we don't have very long lines.
--}
-
--- | Extracts a specific line from a 'Rope.Rope'.
--- Logarithmic in the number of lines.
-extractLine :: Rope.Rope -> Word -> Maybe Rope.Rope
-extractLine rope l = do
-  -- Check for the line being out of bounds
-  let lastLine = Rope.posLine $ Rope.lengthAsPosition rope
-  guard $ l <= lastLine
-
-  let (_, suffix) = Rope.splitAtLine l rope
-      (prefix, _) = Rope.splitAtLine 1 suffix
-  pure prefix
-
--- | Translate a code-point offset into a code-unit offset.
--- Linear in the length of the rope.
-codePointOffsetToCodeUnitOffset :: URope.Rope -> Word -> Maybe Word
-codePointOffsetToCodeUnitOffset rope offset = do
-  -- Check for the position being out of bounds
-  guard $ offset <= URope.length rope
-  -- Split at the given position in *code points*
-  let (prefix, _) = URope.splitAt offset rope
-      -- Convert the prefix to a rope using *code units*
-      utf16Prefix = Rope.fromText $ URope.toText prefix
-      -- Get the length of the prefix in *code units*
-  pure $ Rope.length utf16Prefix
-
--- | Translate a UTF-16 code-unit offset into a code-point offset.
--- Linear in the length of the rope.
-codeUnitOffsetToCodePointOffset :: Rope.Rope -> Word -> Maybe Word
-codeUnitOffsetToCodePointOffset rope offset = do
-  -- Check for the position being out of bounds
-  guard $ offset <= Rope.length rope
-  -- Split at the given position in *code units*
-  (prefix, _) <- Rope.splitAt offset rope
-  -- Convert the prefix to a rope using *code points*
-  let utfPrefix = URope.fromText $ Rope.toText prefix
-      -- Get the length of the prefix in *code points*
-  pure $ URope.length utfPrefix
-
 -- | Given a virtual file, translate a 'CodePointPosition' in that file into a 'J.Position' in that file.
 --
--- Will return 'Nothing' if the requested position is out of bounds of the document.
```

Comment:
What happens now? I'm guessing we get the "clamped" result. I think both behaviours are reasonable but I'd like it to be documented.

Reply:
Yes, it's clamped.